File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-python/scripts/conformance_runner.py
# Copyright (C) 2026 Antony Stubbs and contributors

"""Python's half of the shared cross-language conformance suite (astubbs#242, confluentinc#154).

IT ASSERTS NOTHING, DELIBERATELY. The suite that knows what correct looks like - offset frontiers,
ordering, redelivery, attempt counts - is the Java module
parallel-consumer-proxy-clients/parallel-consumer-proxy-conformance, and it keeps owning that
knowledge for every language. This program's whole job is to DO WHAT THE SCENARIO SAYS and then
exit; if it were free to decide what "correct" means, ten languages would each decide it slightly
differently and the agreement between them would prove nothing.

Its contract - flags, exit codes, the two stdout lines, the behaviour tokens - is documented once,
in parallel-consumer-proxy-clients/parallel-consumer-proxy-conformance/README.md, and is identical
in every language. Read that before changing anything here.

THIS DOES NOT REPLACE THE PACKAGE'S OWN TESTS. The shared suite proves every client behaves
identically on the protocol; tests/ catches what is invisible from outside the process - a channel
inherited across a fork, a worker that dies holding a record, a queue that hands out wrongly. Both
layers are load-bearing.

WHY THE COORDINATION IS multiprocessing PRIMITIVES. This client runs the user's function in worker
PROCESSES, so the scenario's own bookkeeping - the delivery ordinal, "has a second arrived yet",
the ceiling group's held count and generation - cannot be a closure over ordinary variables: each
worker would increment its own copy. They are fork-context primitives, created before the client
exists and inherited by every worker. A threading.Lock or threading.Condition here would coordinate
nothing at all: it would be duplicated by the fork and every worker would hold its own, uncontended.
That is why every prescribed hold below - the barrier included - is written against
multiprocessing.Condition, Event and Value rather than the threading spellings the pseudocode in
the contract reads like.
"""

from __future__ import annotations

import argparse
import multiprocessing
import os
import queue
import sys
import time
from datetime import timedelta
from typing import Any

from parallel_consumer import ClientOptions, ParallelConsumerClient

# Exit statuses ARE the verdict channel. There is no results file, no report message and no second
# protocol: a scenario passed if this process exited 0 and the Java suite's own assertions about
# engine state held.
EXIT_OK = 0
EXIT_BEHAVIOUR_FAILED = 1
EXIT_USAGE = 2

BEHAVIOUR_SUCCEED = "succeed"
BEHAVIOUR_REPORT_NOTHING = "report-nothing"
BEHAVIOUR_FAIL_THEN_SUCCEED = "fail-then-succeed"
BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND = "hold-first-until-second"
BEHAVIOUR_HOLD_UNTIL_CEILING_FULL = "hold-until-ceiling-full"
BEHAVIOURS = (
    BEHAVIOUR_SUCCEED,
    BEHAVIOUR_REPORT_NOTHING,
    BEHAVIOUR_FAIL_THEN_SUCCEED,
    BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND,
    BEHAVIOUR_HOLD_UNTIL_CEILING_FULL,
)

# The exact text a fail-then-succeed run reports. The Java suite asserts the redelivery carries it
# back VERBATIM, so it is a fixed literal of the contract in every language, never composed here.
PRESCRIBED_FAILURE_REASON = "conformance-prescribed-failure"

# Fixed session tunables, contract rather than this runner's judgement: they exist only so scenarios
# converge at unit-test speed against the engine's production defaults (a 5s commit interval, a 1s
# retry delay). Every language sets the same two values.
COMMIT_INTERVAL = timedelta(milliseconds=100)
RETRY_DELAY = timedelta(milliseconds=50)

# How long a report-nothing run keeps its session OPEN after its last observation.
#
# IT IS WHAT MAKES THE NEGATIVE CONTROL A CONTROL. Without it the runner ends the instant the record
# arrives, and a sabotaged runner that DID report success has its report killed in flight by the
# shutdown - so the suite sees an unadvanced offset either way and the scenario passes for a broken
# client. Measured in the Go wave, not reasoned about: reporting success from this behaviour left
# the suite green until the hold existed.
REPORT_NOTHING_HOLD_SECONDS = 3.0

# How long the report-nothing shutdown waits for a record that will never be reported.
#
# THIS IS WHERE PYTHON DEVIATES FROM THE CONTRACT'S LETTER, AND WHY. The contract says a
# report-nothing runner exits "without a clean close" - Go can abandon its session because its
# workers are goroutines that die with the process. Python's workers are PROCESSES, and a runner
# that abandoned them would leave a blocked interpreter behind on the machine for every negative
# control it ran. So the session is shut down instead, with a drain short enough that the held
# record is never reported: observably identical - no outcome for the record, ever - and it reaps
# what it started.
REPORT_NOTHING_DRAIN_SECONDS = 2.0

# How long hold-until-ceiling-full keeps a FULL group held before releasing it.
#
# IT IS WHAT TURNS "the ceiling was never exceeded" FROM A RACE INTO A MEASUREMENT. Release the
# group the instant it fills and a client that declared a larger ceiling still passes - its extra
# records arrive a few milliseconds later, by which time the outstanding count has already fallen
# back. Holding the full ceiling still means the extra dispatch arrives INSIDE the window and prints
# its line while every other record is unresolved. A correct engine cannot dispatch anything during
# the window at all, so the wait costs a conforming client nothing but time.
CEILING_SETTLE_SECONDS = 0.25

DISPATCH_LINE = "dispatch key={key} offset={offset} attempt={attempt} reason={reason}"

# The second line per record, printed the moment the prescribed behaviour has DECIDED that record's
# outcome. Its reason is the failure this runner is REPORTING - empty for a success - not the one
# the record arrived with. The suite reads overlap from the ORDER of the two line types alone, no
# clock involved, which is why both are printed under the delivery lock below.
SETTLED_LINE = "settled key={key} offset={offset} attempt={attempt} reason={reason}"


def main(argv: list[str]) -> int:
    options = parse(argv)
    if isinstance(options, int):
        return options
    return run(options)


def parse(argv: list[str]) -> argparse.Namespace | int:
    """The six flags, spelled identically in every language - including the British --behaviour."""
    parser = argparse.ArgumentParser(prog="conformance-runner", add_help=False)
    parser.add_argument("--scenario", required=True)
    parser.add_argument("--behaviour", required=True, choices=BEHAVIOURS)
    parser.add_argument("--sidecar", required=True)
    parser.add_argument("--expect-dispatches", required=True, type=int)
    parser.add_argument("--max-concurrency", required=True, type=int)
    parser.add_argument("--timeout-seconds", required=True, type=int)
    try:
        parsed = parser.parse_args(argv)
    except SystemExit:
        # argparse exits 2 itself, which is this contract's usage status - but going through
        # SystemExit would skip the flush below, so it is caught and returned.
        return EXIT_USAGE

    if not os.path.isabs(parsed.sidecar):
        print(f"conformance-runner: --sidecar must be absolute, got {parsed.sidecar!r}",
              file=sys.stderr)
        return EXIT_USAGE
    if parsed.expect_dispatches < 1:
        print("conformance-runner: --expect-dispatches must be at least 1", file=sys.stderr)
        return EXIT_USAGE
    if parsed.max_concurrency < 1:
        print("conformance-runner: --max-concurrency must be at least 1", file=sys.stderr)
        return EXIT_USAGE
    if parsed.timeout_seconds < 1:
        print("conformance-runner: --timeout-seconds must be at least 1", file=sys.stderr)
        return EXIT_USAGE
    return parsed


def run(args: argparse.Namespace) -> int:
    deadline = time.monotonic() + args.timeout_seconds
    tracker = Tracker(args.expect_dispatches, args.behaviour, args.max_concurrency,
                      args.timeout_seconds)

    options = ClientOptions(
        # THE SCENARIO NAME IS ALSO THE TOPIC NAME.
        topics=[args.scenario],
        # The ceiling is the SCENARIO's to choose and this runner never derives one - it comes from
        # --max-concurrency and from nothing else. Deriving it from --expect-dispatches, which is
        # what this line used to do, is a ceiling no scenario can reach, so no scenario could ask a
        # client to prove it respected one.
        max_concurrency=args.max_concurrency,
        commit_interval=COMMIT_INTERVAL,
        default_message_retry_delay=RETRY_DELAY,
        # The mock lane builds mock Kafka clients and reads no properties. Real credentials never
        # belong in a conformance test.
        kafka_properties={},
    )
    drain = (REPORT_NOTHING_DRAIN_SECONDS if args.behaviour == BEHAVIOUR_REPORT_NOTHING else 30.0)
    client = ParallelConsumerClient(options, sidecar=args.sidecar, drain_timeout=drain)

    try:
        client.poll(tracker.processor())
    except Exception as failure:
        print(f"conformance-runner: starting the poll: {failure}", file=sys.stderr)
        client.close()
        return EXIT_BEHAVIOUR_FAILED

    # report-nothing completes at OBSERVATION, because by prescription its records are never
    # reported and so can never complete. Every other behaviour completes when the last record it
    # was handed has had its outcome decided.
    if not tracker.await_prescribed_behaviour(deadline):
        print(f"conformance-runner: scenario {args.scenario!r} behaviour "
              f"{args.behaviour!r} did not complete within {args.timeout_seconds}s - observed "
              f"{tracker.observed} of {args.expect_dispatches}, completed {tracker.completed}",
              file=sys.stderr)
        close_quietly(client)
        return EXIT_BEHAVIOUR_FAILED

    if args.behaviour == BEHAVIOUR_REPORT_NOTHING:
        # Hold the session open, reporting nothing, so the suite is watching a LIVE client rather
        # than the wreckage of one - see REPORT_NOTHING_HOLD_SECONDS.
        time.sleep(REPORT_NOTHING_HOLD_SECONDS)
        close_quietly(client)
        return EXIT_OK

    try:
        client.close()
    except Exception as failure:
        print(f"conformance-runner: closing the session: {failure}", file=sys.stderr)
        return EXIT_BEHAVIOUR_FAILED
    return EXIT_OK


def close_quietly(client: ParallelConsumerClient) -> None:
    """Shuts down when the verdict is already decided: a close error must not rewrite it."""
    try:
        client.close()
    except Exception as failure:
        print(f"conformance-runner: while shutting down: {failure}", file=sys.stderr)


def await_ceiling_group(ceiling: Any, held: Any, generation: Any, observed: Any,
                        max_concurrency: int, expected: int, budget: float) -> bool:
    """The cyclic barrier at the heart of hold-until-ceiling-full, ACROSS PROCESSES.

    Block until this record is one of ``max_concurrency`` held at once, keep the full group still
    for CEILING_SETTLE_SECONDS, and release it. A group also releases once every prescribed delivery
    has been observed, so a scenario whose record count is not a multiple of its ceiling cannot
    strand its last, short group.

    Every argument except the last three is a fork-context primitive: the workers holding this
    barrier are different processes, so ``held`` and ``generation`` are shared memory and the
    condition is a multiprocessing one. The threading spellings would compile, run, and synchronize
    each worker with only itself.

    :returns: False if the group never filled inside the budget, which is this runner failing rather
        than the client being wrong about anything.
    """
    deadline = time.monotonic() + budget
    with ceiling:
        mine = generation.value
        held.value += 1
        releasing = held.value >= max_concurrency or observed.value >= expected
        if not releasing:
            while generation.value == mine:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    return False
                ceiling.wait(timeout=remaining)
            return True

    # THE SETTLE WINDOW, HELD OUTSIDE THE LOCK so a record the engine should not be dispatching can
    # still print its arrival if it turns up - that arrival is the whole thing the scenario looks
    # for. A correct engine cannot dispatch anything here, the ceiling being full, so an extra
    # dispatch line inside this window IS the excess.
    time.sleep(CEILING_SETTLE_SECONDS)
    with ceiling:
        held.value = 0
        generation.value += 1
        ceiling.notify_all()
    return True


class Tracker:
    """Counts deliveries and outcomes across the worker processes, and prints the observations.

    It holds no per-record state - only counts - because the client library holds none either, and
    this runner must not become the place where a client's missing bookkeeping is quietly supplied.
    """

    def __init__(self, expected: int, behaviour: str, max_concurrency: int, budget: float) -> None:
        self._expected = expected
        self._behaviour = behaviour
        self._max_concurrency = max_concurrency
        self._budget = budget
        # The fork context, matching the one the client's pool uses: these primitives are created
        # before any worker exists and are inherited by every one of them.
        context = multiprocessing.get_context("fork")
        self._lock = context.Lock()
        self._ordinal = context.Value("i", 0)
        self._second_arrived = context.Event()
        self._held_forever = context.Event()
        # The ceiling group's whole state: how many records are held right now, and which generation
        # of the group they belong to. Both are unlocked shared memory because every access is made
        # under the condition, which carries the only lock there is - the same three lines of state
        # every other language writes out.
        self._ceiling = context.Condition()
        self._held_in_group = context.Value("i", 0, lock=False)
        self._generation = context.Value("l", 0, lock=False)
        self._events: multiprocessing.Queue[str] = context.Queue()
        self.observed = 0
        self.completed = 0

    def processor(self):
        """The user's function, as a closure over this tracker - which is why the pool must fork."""
        expected = self._expected
        behaviour = self._behaviour
        max_concurrency = self._max_concurrency
        budget = self._budget
        lock = self._lock
        ordinal_counter = self._ordinal
        second_arrived = self._second_arrived
        held_forever = self._held_forever
        ceiling = self._ceiling
        held_in_group = self._held_in_group
        generation = self._generation
        events = self._events

        def process(record):
            key = "" if record.key is None else record.key.decode("utf-8", "replace")

            def settled(reason: str) -> None:
                """Prints this record's outcome, under the SAME lock the dispatch line took.

                The suite reads how many records were unresolved at an instant from the running
                difference between the two line types in line order, so the order of the writes has
                to be the order of the events - and here that means the order across processes.
                """
                with lock:
                    print(SETTLED_LINE.format(key=key, offset=record.offset,
                                              attempt=record.attempt, reason=reason), flush=True)

            with lock:
                ordinal_counter.value += 1
                ordinal = ordinal_counter.value
                # Printed at the moment of delivery, before the behaviour acts on it, and under the
                # same lock as the ordinal so the transcript's ORDER is the arrival order: two
                # shards deliver into two processes writing one stdout.
                print(DISPATCH_LINE.format(
                    key=key,
                    offset=record.offset,
                    attempt=record.attempt,
                    reason=record.last_failure_reason or ""), flush=True)
            events.put("observed")
            if ordinal >= 2:
                second_arrived.set()

            if behaviour == BEHAVIOUR_SUCCEED:
                settled("")
                events.put("completed")
                return None

            if behaviour == BEHAVIOUR_REPORT_NOTHING:
                # Never report, and print no settled line EVER: by prescription this record is never
                # resolved, and the absence is the observation. Blocking here is how a Python worker
                # says "this record's function has not returned"; the shutdown terminates the worker
                # with the record still out.
                held_forever.wait()
                return None

            if behaviour == BEHAVIOUR_FAIL_THEN_SUCCEED:
                reason = PRESCRIBED_FAILURE_REASON if record.attempt == 1 else ""
                settled(reason)
                events.put("completed")
                if reason:
                    # A raise IS Python's failure idiom, and its text is the reason verbatim.
                    raise RuntimeError(reason)
                return None

            if behaviour == BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND:
                # Hold the first record until a SECOND is dispatched. Whether one arrives at all,
                # and which key it carries, is the whole of what the scenario is asking - and it is
                # the Java suite that decides what the answer means.
                if ordinal == 1 and not second_arrived.wait(timeout=expected * 60):
                    # The prescription could not be carried out, so this reports a conformance
                    # failure and never completes: the parent's budget is what turns that into
                    # exit 1, which is the path every give-up in this runner already takes.
                    failure = "conformance: no second dispatch arrived while the first was held"
                    settled(failure)
                    raise RuntimeError(failure)
                settled("")
                events.put("completed")
                return None

            if behaviour == BEHAVIOUR_HOLD_UNTIL_CEILING_FULL:
                # Hold EVERY delivery until max_concurrency of them are held at once, keep the full
                # group still for the settle window, then succeed all of them and begin the next
                # group. Blocking is how this layer says the record's function has not returned, so
                # a held record is genuinely unresolved for as long as it looks - which is the
                # property the scenario measures.
                if not await_ceiling_group(ceiling, held_in_group, generation, ordinal_counter,
                                           max_concurrency, expected, budget):
                    failure = f"conformance: the ceiling group of {max_concurrency} never filled"
                    settled(failure)
                    raise RuntimeError(failure)
                settled("")
                events.put("completed")
                return None

            # unreachable: parse() rejects an unknown behaviour before the session opens
            raise RuntimeError(f"conformance: unknown behaviour {behaviour!r}")

        return process

    def await_prescribed_behaviour(self, deadline: float) -> bool:
        """Drains the workers' events until the prescription is finished, or the budget runs out."""
        while True:
            enough = (self.observed if self._behaviour == BEHAVIOUR_REPORT_NOTHING
                      else self.completed)
            if enough >= self._expected:
                return True
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                return False
            try:
                event = self._events.get(timeout=min(remaining, 1.0))
            except queue.Empty:
                # Not yet - the deadline above is what decides when "not yet" becomes "never".
                continue
            if event == "observed":
                self.observed += 1
            elif event == "completed":
                self.completed += 1


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-rust/src/bin/conformance-runner.rs
// Copyright (C) 2026 Antony Stubbs and contributors

//! Rust's half of the shared cross-language conformance suite (astubbs#242, confluentinc#154).
//!
//! IT ASSERTS NOTHING, DELIBERATELY. The suite that knows what correct looks like - offset
//! frontiers, ordering, redelivery, attempt counts - is the Java module
//! `parallel-consumer-proxy-clients/parallel-consumer-proxy-conformance`, and it keeps owning that
//! knowledge for every language. This binary's whole job is to DO WHAT THE SCENARIO SAYS and then
//! exit; if it were free to decide what "correct" means, ten languages would each decide it
//! slightly differently and the agreement between them would prove nothing.
//!
//! Its contract - flags, exit codes, the two stdout lines per record, the behaviour tokens - is
//! documented once, in that module's `README.md`, and is identical in every language.
//!
//! THIS DOES NOT REPLACE THE CRATE'S OWN TESTS. The shared suite proves every client behaves
//! identically on the protocol; `tests/` catches what is invisible from outside the process - a
//! blocking call inside an executor, a task that outlives its session, a child that is never
//! reaped. Both layers are load-bearing.
//!
//! A CURRENT-THREAD RUNTIME, ON PURPOSE. A bin target sees only the crate's own dependencies, and
//! `rt-multi-thread` is a dev-dependency here - but the stricter runtime is the better test anyway:
//! every executor, the transport task and this runner share one thread, so a client that blocked
//! instead of awaiting would deadlock the concurrency scenario rather than getting away with it on
//! a spare core.

use std::collections::HashMap;
use std::path::Path;
use std::process::ExitCode;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use parallel_consumer_proxy_client::{
    ClientOptions, InboundRecord, Outcome, ParallelConsumerClient, ProcessingError,
};
use tokio::sync::watch;
use tokio::time::Instant;

/// Exit statuses ARE the verdict channel. There is no results file and no report message: a
/// scenario passed if this process exited 0 and the Java suite's own assertions about engine state
/// held.
const EXIT_OK: u8 = 0;
const EXIT_BEHAVIOUR_FAILED: u8 = 1;
const EXIT_USAGE: u8 = 2;

const BEHAVIOUR_SUCCEED: &str = "succeed";
const BEHAVIOUR_REPORT_NOTHING: &str = "report-nothing";
const BEHAVIOUR_FAIL_THEN_SUCCEED: &str = "fail-then-succeed";
const BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND: &str = "hold-first-until-second";
const BEHAVIOUR_HOLD_UNTIL_CEILING_FULL: &str = "hold-until-ceiling-full";

/// The exact text a fail-then-succeed run reports. The Java suite asserts the redelivery carries it
/// back VERBATIM, so it is a fixed literal of the contract in every language, never composed here.
const PRESCRIBED_FAILURE_REASON: &str = "conformance-prescribed-failure";

// Fixed session tunables, contract rather than this runner's judgement: they exist only so
// scenarios converge at unit-test speed against the engine's production defaults (a 5s commit
// interval, a 1s retry delay). Every language sets the same two values.
const COMMIT_INTERVAL: Duration = Duration::from_millis(100);
const RETRY_DELAY: Duration = Duration::from_millis(50);

/// How long a report-nothing run keeps its session OPEN after its last observation.
///
/// IT IS WHAT MAKES THE NEGATIVE CONTROL A CONTROL. Without it the runner exits the instant the
/// record arrives, and a sabotaged runner that DID report success has its report killed in flight
/// by the process exit - so the suite sees an unadvanced offset either way and the scenario passes
/// for a broken client. Measured in the Go wave, not reasoned about: reporting success from this
/// behaviour left the suite green until the hold existed.
const REPORT_NOTHING_HOLD: Duration = Duration::from_secs(3);

/// How long `hold-until-ceiling-full` keeps a FULL group held before releasing it.
///
/// IT IS WHAT TURNS "THE CEILING WAS NEVER EXCEEDED" FROM A RACE INTO A MEASUREMENT. Release the
/// group the instant it fills and a client that declared a larger ceiling still passes - its extra
/// records arrive a few milliseconds later, by which time the outstanding count has already fallen
/// back. Holding the full ceiling still means the extra dispatch arrives INSIDE the window and
/// prints its line while every other record is unresolved. A correct engine cannot dispatch
/// anything during the window at all, so the wait costs a conforming client nothing but time.
const CEILING_SETTLE: Duration = Duration::from_millis(250);

#[tokio::main(flavor = "current_thread")]
async fn main() -> ExitCode {
    let arguments = match Arguments::parse(std::env::args().skip(1).collect()) {
        Ok(arguments) => arguments,
        Err(problem) => {
            eprintln!("conformance-runner: {problem}");
            return ExitCode::from(EXIT_USAGE);
        }
    };
    ExitCode::from(run(arguments).await)
}

struct Arguments {
    scenario: String,
    behaviour: String,
    sidecar: String,
    expect_dispatches: usize,
    max_concurrency: usize,
    timeout_seconds: u64,
}

impl Arguments {
    /// The six flags, spelled identically in every language - including the British `--behaviour`.
    fn parse(argv: Vec<String>) -> Result<Self, String> {
        let mut values: HashMap<String, String> = HashMap::new();
        let mut index = 0;
        while index < argv.len() {
            let flag = &argv[index];
            let value = argv
                .get(index + 1)
                .ok_or_else(|| format!("{flag} takes a value"))?;
            if !flag.starts_with("--") {
                return Err(format!("expected --flag value pairs, got {flag}"));
            }
            values.insert(flag.clone(), value.clone());
            index += 2;
        }

        let take = |name: &str| -> Result<String, String> {
            values
                .get(name)
                .filter(|value| !value.is_empty())
                .cloned()
                .ok_or_else(|| format!("{name} is required"))
        };

        let scenario = take("--scenario")?;
        let behaviour = take("--behaviour")?;
        let sidecar = take("--sidecar")?;
        if !matches!(
            behaviour.as_str(),
            BEHAVIOUR_SUCCEED
                | BEHAVIOUR_REPORT_NOTHING
                | BEHAVIOUR_FAIL_THEN_SUCCEED
                | BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND
                | BEHAVIOUR_HOLD_UNTIL_CEILING_FULL
        ) {
            return Err(format!("unknown behaviour {behaviour:?}"));
        }
        if !Path::new(&sidecar).is_absolute() {
            return Err(format!("--sidecar must be absolute, got {sidecar:?}"));
        }
        let expect_dispatches: usize = take("--expect-dispatches")?
            .parse()
            .map_err(|_| "--expect-dispatches must be a positive integer".to_owned())?;
        let max_concurrency: usize = take("--max-concurrency")?
            .parse()
            .map_err(|_| "--max-concurrency must be a positive integer".to_owned())?;
        let timeout_seconds: u64 = take("--timeout-seconds")?
            .parse()
            .map_err(|_| "--timeout-seconds must be a positive integer".to_owned())?;
        if expect_dispatches < 1 {
            return Err("--expect-dispatches must be at least 1".to_owned());
        }
        if max_concurrency < 1 {
            return Err("--max-concurrency must be at least 1".to_owned());
        }
        if timeout_seconds < 1 {
            return Err("--timeout-seconds must be at least 1".to_owned());
        }

        Ok(Self { scenario, behaviour, sidecar, expect_dispatches, max_concurrency, timeout_seconds })
    }
}

async fn run(arguments: Arguments) -> u8 {
    let budget = Duration::from_secs(arguments.timeout_seconds);
    let tracker = Arc::new(Tracker::new(arguments.expect_dispatches, arguments.max_concurrency));

    let client = match ParallelConsumerClient::connect(ClientOptions {
        sidecar_path: arguments.sidecar.clone().into(),
        // THE SCENARIO NAME IS ALSO THE TOPIC NAME.
        topics: vec![arguments.scenario.clone()],
        // THE CEILING IS THE SCENARIO'S TO CHOOSE, and this runner never derives one: it is set
        // from --max-concurrency and from nothing else. Deriving it from --expect-dispatches, which
        // is what this line used to do, is by construction a ceiling no scenario can reach.
        max_concurrency: Some(arguments.max_concurrency as i32),
        commit_interval: Some(COMMIT_INTERVAL),
        default_message_retry_delay: Some(RETRY_DELAY),
        // The mock lane builds mock Kafka clients and reads no properties. Real credentials never
        // belong in a conformance test.
        kafka_properties: HashMap::new(),
        instance_tag: Some("conformance-runner-rust".to_owned()),
        ..Default::default()
    })
    .await
    {
        Ok(client) => client,
        Err(error) => {
            eprintln!("conformance-runner: opening the session: {error}");
            return EXIT_BEHAVIOUR_FAILED;
        }
    };

    let behaviour = arguments.behaviour.clone();
    let processor_tracker = Arc::clone(&tracker);
    if let Err(error) = client.poll(move |record: InboundRecord| {
        let tracker = Arc::clone(&processor_tracker);
        let behaviour = behaviour.clone();
        async move { process(&behaviour, &tracker, budget, record).await }
    }) {
        eprintln!("conformance-runner: starting the poll: {error}");
        let _ = client.shutdown().await;
        return EXIT_BEHAVIOUR_FAILED;
    }

    // report-nothing completes at OBSERVATION, because by prescription its records are never
    // reported and so can never complete. Every other behaviour completes when the last record it
    // was handed has had its outcome decided.
    let report_nothing = arguments.behaviour == BEHAVIOUR_REPORT_NOTHING;
    let finished = tracker.await_prescribed_behaviour(report_nothing, budget).await;
    if !finished {
        eprintln!(
            "conformance-runner: scenario {:?} behaviour {:?} did not complete within {}s - \
             observed {} of {}, completed {}",
            arguments.scenario,
            arguments.behaviour,
            arguments.timeout_seconds,
            tracker.observed(),
            arguments.expect_dispatches,
            tracker.completed()
        );
        let _ = client.shutdown().await;
        return EXIT_BEHAVIOUR_FAILED;
    }

    if report_nothing {
        // Hold the session open, reporting nothing, so the suite is watching a LIVE client rather
        // than the wreckage of one - see REPORT_NOTHING_HOLD.
        tokio::time::sleep(REPORT_NOTHING_HOLD).await;
        // PRESCRIBED: the record is never reported and the session is abandoned rather than
        // drained - a worker that vanished mid-record is exactly what this scenario models. Exiting
        // closes the sidecar's lifecycle pipe, which reaps it, so nothing is leaked by not closing.
        return EXIT_OK;
    }

    if let Err(error) = client.shutdown().await {
        eprintln!("conformance-runner: closing the session: {error}");
        return EXIT_BEHAVIOUR_FAILED;
    }
    EXIT_OK
}

async fn process(
    behaviour: &str,
    tracker: &Tracker,
    budget: Duration,
    record: InboundRecord,
) -> Result<Outcome, ProcessingError> {
    let attempt = record.attempt;
    let ordinal = tracker.observe(&record);

    match behaviour {
        BEHAVIOUR_SUCCEED => {
            tracker.complete();
            tracker.settle(&record, None)
        }

        BEHAVIOUR_REPORT_NOTHING => {
            // Never report, and PRINT NO settled LINE - by prescription this record is never
            // resolved, and the absence of the line is the observation. A future that never
            // resolves is how a Rust worker says "this record's function has not returned"; the
            // process exits with the record still in flight.
            std::future::pending::<()>().await;
            unreachable!("a pending future never resolves")
        }

        BEHAVIOUR_FAIL_THEN_SUCCEED => {
            tracker.complete();
            if attempt == 1 {
                // `Err` IS the language's failure idiom, and its reason travels verbatim.
                return tracker.settle(&record, Some(PRESCRIBED_FAILURE_REASON));
            }
            tracker.settle(&record, None)
        }

        BEHAVIOUR_HOLD_FIRST_UNTIL_SECOND => {
            if ordinal == 1 {
                // Hold the first record until a SECOND is dispatched. Whether one arrives at all,
                // and which key it carries, is the whole of what the scenario is asking - and it is
                // the Java suite that decides what the answer means.
                tracker.second_arrived().await;
            }
            tracker.complete();
            tracker.settle(&record, None)
        }

        BEHAVIOUR_HOLD_UNTIL_CEILING_FULL => {
            // Hold until --max-concurrency records are held AT ONCE, keep the full group still for
            // CEILING_SETTLE, then release the whole group and start the next one.
            if !tracker.enter_ceiling_group(budget).await {
                // The prescription could not be carried out, so the run failed: the reason is
                // reported and printed, and the top-level budget in await_prescribed_behaviour -
                // this runner's existing route to exit 1 - is left to expire, because this record
                // is deliberately never counted as completed.
                let never_filled =
                    format!("conformance: the ceiling group of {} never filled", tracker.max_concurrency);
                return tracker.settle(&record, Some(&never_filled));
            }
            tracker.complete();
            tracker.settle(&record, None)
        }

        // unreachable: Arguments::parse rejects an unknown behaviour before the session opens
        other => Err(ProcessingError::new(format!("conformance: unknown behaviour {other:?}"))),
    }
}

/// Counts deliveries and outcomes, prints both observation lines, and holds the ceiling group's
/// barrier - the only three things a runner has to keep. It holds no per-record state -
/// only counts - because the client library holds none either, and this runner must not become the
/// place where a client's missing bookkeeping is quietly supplied.
struct Tracker {
    expected: usize,
    max_concurrency: usize,
    observed: AtomicUsize,
    completed: AtomicUsize,
    /// Taken for the increment AND the print together, and for the `settled` print as well, so the
    /// transcript's order is the order the events happened in - which is the whole of what the
    /// suite reads overlap from.
    printing: Mutex<()>,
    counts: watch::Sender<(usize, usize)>,
    /// How many records the `hold-until-ceiling-full` barrier is holding right now. It guards the
    /// generation below too: every read and every bump of the generation happens under this lock.
    held: Mutex<usize>,
    /// Which generation of the ceiling group a held record belongs to - the barrier's release
    /// signal, bumped once per group.
    ///
    /// A WATCH CHANNEL RATHER THAN A `Condvar`, because every wait here happens inside an executor
    /// task on a current-thread runtime: a `Condvar` would park the one thread that still has to
    /// dispatch the rest of the group, and the barrier could never fill. The same reasoning makes
    /// the settle window a `tokio::time::sleep` rather than a `thread::sleep`.
    generation: watch::Sender<u64>,
}

impl Tracker {
    fn new(expected: usize, max_concurrency: usize) -> Self {
        Self {
            expected,
            max_concurrency,
            observed: AtomicUsize::new(0),
            completed: AtomicUsize::new(0),
            printing: Mutex::new(()),
            counts: watch::channel((0, 0)).0,
            held: Mutex::new(0),
            generation: watch::channel(0).0,
        }
    }

    /// Prints the delivery and returns its 1-based ordinal in arrival order.
    fn observe(&self, record: &InboundRecord) -> usize {
        let ordinal = {
            let _printing = self.printing.lock().expect("the print lock is never poisoned");
            let ordinal = self.observed.fetch_add(1, Ordering::SeqCst) + 1;
            // Printed at the moment of delivery, before the behaviour acts on it. reason comes last
            // because it is worker-supplied text that may contain spaces.
            println!(
                "dispatch key={} offset={} attempt={} reason={}",
                record.key.as_deref().map(String::from_utf8_lossy).unwrap_or_default(),
                record.offset,
                record.attempt,
                record.last_failure_reason.as_deref().unwrap_or("")
            );
            ordinal
        };
        self.publish();
        ordinal
    }

    /// Prints the `settled` line the moment this record's outcome has been DECIDED, and returns
    /// that outcome so the caller reports exactly what it printed.
    ///
    /// A dispatch line opens a record's unresolved window and this one closes it, so the running
    /// difference between the two - read in line order - is how many records this client was
    /// holding at that instant. That is why it prints under the same lock the dispatch side takes.
    fn settle(&self, record: &InboundRecord, failure: Option<&str>) -> Result<Outcome, ProcessingError> {
        {
            let _printing = self.printing.lock().expect("the print lock is never poisoned");
            println!(
                "settled key={} offset={} attempt={} reason={}",
                record.key.as_deref().map(String::from_utf8_lossy).unwrap_or_default(),
                record.offset,
                record.attempt,
                failure.unwrap_or("")
            );
        }
        match failure {
            // `Err` IS the language's failure idiom, and its reason travels verbatim.
            Some(reason) => Err(ProcessingError::new(reason)),
            None => Ok(Outcome::success()),
        }
    }

    /// The cyclic barrier at the heart of `hold-until-ceiling-full`: hold until this record is one
    /// of `max_concurrency` held at once, keep the full group still for [`CEILING_SETTLE`], then
    /// release it.
    ///
    /// A group also releases once every prescribed delivery has been observed, so a scenario whose
    /// record count is not a multiple of its ceiling cannot strand its last, short group.
    ///
    /// AWAITING IS HOW A RUST WORKER SAYS THE RECORD IS STILL OUT: the future has not resolved, so
    /// the record is unresolved, while the executor task is parked rather than the runtime blocked.
    ///
    /// Returns false if the group never filled inside the budget, which is this runner failing
    /// rather than the client being wrong about anything.
    async fn enter_ceiling_group(&self, budget: Duration) -> bool {
        let deadline = Instant::now() + budget;
        let mut generations = self.generation.subscribe();

        let my_generation = {
            let mut held = self.held.lock().expect("the ceiling lock is never poisoned");
            *held += 1;
            let releasing = *held >= self.max_concurrency || self.observed() >= self.expected;
            if releasing {
                None
            } else {
                Some(*self.generation.borrow())
            }
        };

        let Some(my_generation) = my_generation else {
            // THE SETTLE WINDOW, HELD OUTSIDE THE LOCK so a record the engine should not be
            // dispatching can still print its arrival if it turns up. A correct engine can dispatch
            // nothing here - the ceiling is full - so an extra line inside this window IS the
            // excess this scenario looks for.
            tokio::time::sleep(CEILING_SETTLE).await;
            let mut held = self.held.lock().expect("the ceiling lock is never poisoned");
            *held = 0;
            self.generation.send_modify(|generation| *generation += 1);
            return true;
        };

        loop {
            // The generation's VALUE, not merely a wakeup: a bump that landed between the unlock
            // and the first look is already visible here, so it cannot be missed.
            if *generations.borrow_and_update() != my_generation {
                return true;
            }
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return false;
            }
            match tokio::time::timeout(remaining, generations.changed()).await {
                Ok(Ok(())) => {}
                // The sender lives in this tracker and outlives every executor, so a closed channel
                // cannot happen; the budget running out is the real case, and either way this group
                // did not fill.
                Ok(Err(_)) | Err(_) => return false,
            }
        }
    }

    fn complete(&self) {
        self.completed.fetch_add(1, Ordering::SeqCst);
        self.publish();
    }

    fn observed(&self) -> usize {
        self.observed.load(Ordering::SeqCst)
    }

    fn completed(&self) -> usize {
        self.completed.load(Ordering::SeqCst)
    }

    fn publish(&self) {
        let _ = self.counts.send((self.observed(), self.completed()));
    }

    /// Resolves once a second delivery has been observed - the instrument the ordering scenario is.
    async fn second_arrived(&self) {
        let mut counts = self.counts.subscribe();
        while counts.borrow().0 < 2 {
            if counts.changed().await.is_err() {
                return;
            }
        }
    }

    /// Whether the prescription finished inside the budget.
    async fn await_prescribed_behaviour(&self, at_observation: bool, budget: Duration) -> bool {
        let expected = self.expected;
        let mut counts = self.counts.subscribe();
        let enough = |(observed, completed): (usize, usize)| {
            if at_observation { observed >= expected } else { completed >= expected }
        };
        tokio::time::timeout(budget, async {
            while !enough(*counts.borrow()) {
                if counts.changed().await.is_err() {
                    return;
                }
            }
        })
        .await
        .is_ok()
            && enough((self.observed(), self.completed()))
    }
}
